File: frontend/lib/features/ai/domain/unified_assistant.dart
class UnifiedAssistantClaim {
  const UnifiedAssistantClaim({
    required this.claimId,
    required this.claimClass,
    required this.text,
    required this.sourceRefs,
    this.toolRefs = const [],
    this.estimateStatus,
    this.confidence,
    this.assumptions = const [],
    this.missingInputs = const [],
    this.confirmOrRefute,
  });
  factory UnifiedAssistantClaim.fromJson(Map<String, dynamic> json) =>
      UnifiedAssistantClaim(
        claimId: json['claim_id'] as String,
        claimClass: json['claim_class'] as String,
        text: json['text'] as String,
        sourceRefs: (json['source_refs'] as List<dynamic>? ?? const [])
            .cast<String>(),
        toolRefs: (json['tool_refs'] as List<dynamic>? ?? const [])
            .cast<String>(),
        estimateStatus: json['estimate_status'] as String?,
        confidence: json['confidence'] as String?,
        assumptions: (json['assumptions'] as List<dynamic>? ?? const [])
            .cast<String>(),
        missingInputs: (json['missing_inputs'] as List<dynamic>? ?? const [])
            .cast<String>(),
        confirmOrRefute: json['confirm_or_refute'] as String?,
      );
  final String claimId;
  final String claimClass;
  final String text;
  final List<String> sourceRefs;
  final List<String> toolRefs;
  final String? estimateStatus;
  final String? confidence;
  final List<String> assumptions;
  final List<String> missingInputs;
  final String? confirmOrRefute;
}

class UnifiedAssistantSource {
  const UnifiedAssistantSource({
    required this.sourceRef,
    required this.sourceType,
    required this.title,
    required this.excerpt,
    required this.whyUsed,
    required this.supportsClaimIds,
    this.sourceId,
    this.route,
    this.externalAnalysis = false,
  });
  factory UnifiedAssistantSource.fromJson(Map<String, dynamic> json) =>
      UnifiedAssistantSource(
        sourceRef: json['source_ref'] as String,
        sourceType: json['source_type'] as String,
        sourceId: (json['source_id'] as num?)?.toInt(),
        title: json['title'] as String,
        excerpt: json['excerpt'] as String,
        whyUsed: json['why_used'] as String,
        supportsClaimIds:
            (json['supports_claim_ids'] as List<dynamic>? ?? const [])
                .cast<String>(),
        route: json['route'] as String?,
        externalAnalysis: json['external_analysis'] as bool? ?? false,
      );
  final String sourceRef;
  final String sourceType;
  final int? sourceId;
  final String title;
  final String excerpt;
  final String whyUsed;
  final List<String> supportsClaimIds;
  final String? route;
  final bool externalAnalysis;
}

class UnifiedAssistantAnswer {
  const UnifiedAssistantAnswer({
    required this.requestId,
    required this.answer,
    required this.status,
    required this.progress,
    required this.targetScope,
    required this.claims,
    required this.sources,
    required this.usedTools,
    required this.externalAnalysisUsed,
    this.model,
    this.errorMessage,
  });
  factory UnifiedAssistantAnswer.fromJson(Map<String, dynamic> json) =>
      UnifiedAssistantAnswer(
        requestId: json['request_id'] as String,
        answer: json['answer'] as String,
        status: json['status'] as String,
        progress: json['progress'] as String,
        targetScope: json['target_scope'] as String,
        claims: (json['claims'] as List<dynamic>? ?? const [])
            .map(
              (item) => UnifiedAssistantClaim.fromJson(
                Map<String, dynamic>.from(item as Map),
              ),
            )
            .toList(growable: false),
        sources: (json['sources'] as List<dynamic>? ?? const [])
            .map(
              (item) => UnifiedAssistantSource.fromJson(
                Map<String, dynamic>.from(item as Map),
              ),
            )
            .toList(growable: false),
        usedTools: (json['used_tools'] as List<dynamic>? ?? const [])
            .cast<String>(),
        externalAnalysisUsed: json['external_analysis_used'] as bool? ?? false,
        model: json['model'] as String?,
        errorMessage: json['error_message'] as String?,
      );
  bool get isPending =>
      status == 'advanced_queued' || status == 'advanced_processing';
  final String requestId;
  final String answer;
  final String status;
  final String progress;
  final String targetScope;
  final List<UnifiedAssistantClaim> claims;
  final List<UnifiedAssistantSource> sources;
  final List<String> usedTools;
  final bool externalAnalysisUsed;
  final String? model;
  final String? errorMessage;
}

File: backend/app/schemas/unified_assistant.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class UnifiedConversationMessage(BaseModel):
    model_config = ConfigDict(extra="forbid")
    role: Literal["user", "assistant"]
    content: str = Field(min_length=1, max_length=2000)

    @field_validator("content")
    @classmethod
    def normalize_content(cls, value: str) -> str:
        return " ".join(value.split())


class UnifiedAssistantRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    question: str = Field(min_length=2, max_length=2000)
    client_id: int | None = Field(default=None, gt=0)
    candidate_id: int | None = Field(default=None, gt=0)
    document_id: int | None = Field(default=None, gt=0)
    mail_source_id: int | None = Field(default=None, gt=0)
    inspection_id: int | None = Field(default=None, gt=0)
    attempt_id: str | None = Field(default=None, min_length=8, max_length=80, pattern=r"^[A-Za-z0-9_-]+$")
    conversation: list[UnifiedConversationMessage] = Field(default_factory=list, max_length=8)

    @field_validator("question")
    @classmethod
    def normalize_question(cls, value: str) -> str:
        return " ".join(value.split())

    @model_validator(mode="after")
    def bound_context(self):
        if sum(len(item.content) for item in self.conversation) > 6000:
            raise ValueError("Conversation history is too large")
        return self


class UnifiedClaim(BaseModel):
    model_config = ConfigDict(extra="forbid")
    claim_id: str
    claim_class: Literal["FACT", "ESTIMATE", "HYPOTHESIS", "MISSING"]
    text: str
    source_refs: list[str] = Field(default_factory=list)
    tool_refs: list[str] = Field(default_factory=list)
    estimate_status: Literal["ESTIMABLE", "NOT_ESTIMABLE"] | None = None
    confidence: Literal["HIGH", "MEDIUM", "LOW"] | None = None
    assumptions: list[str] = Field(default_factory=list)
    missing_inputs: list[str] = Field(default_factory=list)
    confirm_or_refute: str | None = None


class UnifiedSource(BaseModel):
    model_config = ConfigDict(extra="forbid")
    source_ref: str
    source_type: str
    source_id: int | None = None
    title: str
    excerpt: str
    why_used: str
    supports_claim_ids: list[str]
    route: str | None = None
    external_analysis: bool = False


class UnifiedAssistantResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    request_id: str
    answer: str
    status: Literal[
        "accepted_local", "advanced_queued", "advanced_processing",
        "accepted_advanced", "review_required", "failed", "timed_out", "cancelled",
        "document_preparation_queued", "document_preparation_running", "resume_queued",
    ]
    progress: Literal[
        "collecting", "preparing_document", "analyzing", "advanced_analysis", "validating", "complete"
    ]
    target_scope: str
    claims: list[UnifiedClaim]
    sources: list[UnifiedSource]
    used_tools: list[str]
    model: str | None = None
    external_analysis_used: bool = False
    error_message: str | None = None
    current_stage: str | None = None
    last_progress_at: str | None = None
    can_cancel: bool = False
    delayed: bool = False
